tests/test_resolver.py

```diff
@@ -72,6 +75,35 @@ def test_provider_choose_sdist():
         assert str(candidate.version) == "1.3.2"
 
 
+def test_provider_choose_either_with_constraint():
```

Comment:
We have some examples in this test file where we mock out the return value from the package server to avoid having to have network access to run the unit tests. That also makes the tests more reliable, since they don't rely on the remote service returning exactly the same answer each time.

Reply:
Well, I see you're doing that below, just not in the method definition, so never mind.